File: twoSum.py
from typing import List


def twoNumberSum(array, targetSum):
    hashMap = {}
    for idx in range(len(array)):
        currentNum = array[idx]
        difference = targetSum-currentNum
        if difference not in hashMap and currentNum not in hashMap:
            hashMap[difference] = currentNum
        else:
            return [currentNum, hashMap[currentNum]]
    return []


def twoNumberSumV2(array, targetSum):
    for idx in range(len(array)-1):
        firstNum = array[idx]
        for idx2 in range(idx+1,len(array)):
            secondNum = array[idx2]
            if secondNum + firstNum  == targetSum :
                return [firstNum, secondNum]
            
    return []


def twoNumberSumV4(array:List[int], targetSum:int)->List[int]:
    leftPointer = 0
    rightPointer = len(array) - 1
    
    array.sort()
    while leftPointer < rightPointer:
        currentSum = array[leftPointer] + array[rightPointer]
        if currentSum > targetSum :
            rightPointer -= 1
        elif currentSum < targetSum :
            leftPointer += 1
        else :
            return [array[leftPointer], array[rightPointer]]
    return  []

File: twoSum.go
package main

import "sort"

func TwoNumberSum(array []int, target int) []int {
	result := []int{}
	hash := make(map[int]int)
	for _, curr := range array {
		difference := target - curr
		_, diffOk := hash[difference]
		_, currOk := hash[curr]
		if !diffOk && !currOk {
			hash[difference] = curr
		} else {
			result = append(result, curr, hash[curr])
		}
	}
	return result
}

func TwoNumberSumV3(array []int, target int) []int {
	result := []int{}
	hash := make(map[int]struct{})
	for _, curr := range array {
		difference := target - curr
		_, diffOk := hash[difference]
		if !diffOk {
			hash[curr] = struct{}{}
		} else {
			result = append(result, curr, difference)
		}
	}
	return result
}

func TwoNumberSumV2(array []int, target int) []int {
	for idx := 0; idx < len(array); idx++ {
		firstNum := array[idx]
		for idx2 := 0; idx2 < len(array); idx2++ {
			secondNum := array[idx2]
			if firstNum+secondNum == target {
				return []int{firstNum, secondNum}
			}
		}
	}
	return []int{}
}

func TwoNumberSumV4(array []int, target int) []int {
	leftPointer := 0
	rightPointer := len(array) - 1

	sort.Ints(array)

	for leftPointer < rightPointer {
		currentSum := array[leftPointer] + array[rightPointer]
		if currentSum > target {
			rightPointer -= 1
		} else if currentSum < target {
			leftPointer += 1
		} else {
			return []int{array[leftPointer], array[rightPointer]}
		}
	}
	return []int{}
}
